File: ios/Classes/TripProgressOverlay.swift
import UIKit
import AudioToolbox

/// Displays a trip progress overlay above the navigation info panel.
///
/// Shows:
/// - Skip prev/next buttons (configurable)
/// - Icon + Next waypoint name + Distance/time to it
/// - Progress bar + "Waypoint X/Y" + Total distance remaining
/// - ETA
/// - End navigation button (configurable)
///
/// The appearance can be customized via [TripProgressConfig] and [TripProgressTheme].
///
/// Layout:
/// ```
/// ┌──────────────────────────────────────────────────────────┐
/// │  [◀]   [icon] Waypoint Name                      [▶]    │
/// │              2.3 mi • ~4 min                             │
/// │  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   │
/// │  Waypoint 3/8                      45 mi remaining      │
/// │                    ETA 2:45pm                           │
/// │                                                         │
/// │                   [End Navigation]                      │
/// └──────────────────────────────────────────────────────────┘
/// ```
public class TripProgressOverlay {

    private weak var parentViewController: UIViewController?
    private var progressCard: UIView?
    private var isVisible = false

    // Configuration
    private var config: TripProgressConfig
    private var iconProvider: IconProvider

    // Callbacks
    public var onSkipPrevious: (() -> Void)?
    public var onSkipNext: (() -> Void)?
    public var onEndNavigation: (() -> Void)?

    // UI elements for updating
    private var iconView: UIImageView?
    private var iconContainer: UIView?
    private var waypointNameLabel: UILabel?
    private var distanceTimeLabel: UILabel?
    private var progressView: UIProgressView?
    private var progressTextLabel: UILabel?
    private var totalDistanceLabel: UILabel?
    private var etaLabel: UILabel?
    private var prevButton: UIButton?
    private var nextButton: UIButton?
    private var endNavButton: UIButton?

    // Current state for button enable/disable
    private var currentWaypointIndex = 0
    private var totalWaypoints = 0

    // Bottom margin to position above the info panel
    private let bottomMargin: CGFloat = 180

    // Theme shortcut
    private var theme: TripProgressTheme { config.theme }

    public init(
        parentViewController: UIViewController,
        config: TripProgressConfig = .defaults(),
        iconProvider: IconProvider = DefaultIconProvider.shared
    ) {
        self.parentViewController = parentViewController
        self.config = config
        self.iconProvider = iconProvider
    }

    /// Update the configuration.
    public func setConfig(_ config: TripProgressConfig) {
        self.config = config
        // If visible, recreate the card with new config
        if isVisible {
            hide(animated: false)
            show()
        }
    }

    /// Show the trip progress overlay.
    public func show() {
        guard !isVisible, let parentVC = parentViewController else { return }

        // Create the progress card
        let card = createProgressCard()
        progressCard = card

        // Add to parent view
        parentVC.view.addSubview(card)

        // Set up constraints
        card.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            card.leadingAnchor.constraint(equalTo: parentVC.view.leadingAnchor, constant: 12),
            card.trailingAnchor.constraint(equalTo: parentVC.view.trailingAnchor, constant: -12),
            card.bottomAnchor.constraint(equalTo: parentVC.view.safeAreaLayoutGuide.bottomAnchor, constant: -bottomMargin)
        ])

        // Animate in
        card.alpha = 0
        card.transform = CGAffineTransform(translationX: 0, y: 30)

        UIView.animate(withDuration: 0.3, delay: 0, options: .curveEaseOut) {
            card.alpha = 1
            card.transform = .identity
        }

        isVisible = true
        print("TripProgressOverlay: Shown")
    }

    /// Hide the trip progress overlay.
    public func hide(animated: Bool = true) {
        guard let card = progressCard else {
            progressCard = nil
            isVisible = false
            return
        }

        if animated {
            UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseIn) {
                card.alpha = 0
                card.transform = CGAffineTransform(translationX: 0, y: 30)
            } completion: { [weak self] _ in
                card.removeFromSuperview()
                self?.progressCard = nil
                self?.isVisible = false
            }
        } else {
            card.removeFromSuperview()
            progressCard = nil
            isVisible = false
        }
    }

    /// Update the overlay with new progress data.
    public func updateProgress(_ data: TripProgressData) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            self.currentWaypointIndex = data.currentWaypointIndex
            self.totalWaypoints = data.totalWaypoints

            // Update icon
            self.iconView?.image = self.iconProvider.getIcon(iconId: data.nextWaypointIconId, category: data.nextWaypointCategory)
            self.iconContainer?.backgroundColor = self.iconProvider.getCategoryColor(data.nextWaypointCategory, theme: self.theme)

            // Update waypoint name
            var displayName = data.nextWaypointName
            if displayName.count > 22 {
                displayName = String(displayName.prefix(19)) + "..."
            }
            self.waypointNameLabel?.text = displayName

            // Update distance and time
            if self.config.showDistanceToNext || self.config.showDurationToNext {
                let distStr = self.config.showDistanceToNext ? data.getFormattedDistanceToNext() : ""
                let timeStr = self.config.showDurationToNext ? data.getFormattedDurationToNext() : ""
                let separator = (self.config.showDistanceToNext && self.config.showDurationToNext) ? " • " : ""
                self.distanceTimeLabel?.text = "\(distStr)\(separator)\(timeStr)"
            }

            // Update progress bar
            if self.config.showProgressBar {
                self.progressView?.progress = data.progressFraction
            }

            // Update progress text
            if self.config.showWaypointCount {
                self.progressTextLabel?.text = data.progressString
            }

            // Update total distance
            if self.config.showTotalDistance {
                self.totalDistanceLabel?.text = "\(data.getFormattedTotalDistanceRemaining()) remaining"
            }

            // Update ETA
            if self.config.showEta {
                self.etaLabel?.text = "ETA \(data.getFormattedEta())"
            }

            // Update button states
            self.updateButtonStates()
        }
    }

    private func updateButtonStates() {
        guard config.showSkipButtons else { return }

        // Prev button: disabled if at first waypoint
        let canGoPrev = currentWaypointIndex > 0
        prevButton?.alpha = canGoPrev ? 1.0 : 0.3
        prevButton?.isEnabled = canGoPrev

        // Next button: disabled if at last waypoint
        let canGoNext = currentWaypointIndex < totalWaypoints - 1
        nextButton?.alpha = canGoNext ? 1.0 : 0.3
        nextButton?.isEnabled = canGoNext
    }

    private func playButtonSound() {
        guard config.enableAudioFeedback else { return }
        AudioServicesPlaySystemSound(1104) // Tock sound
    }

    @objc private func prevButtonTapped() {
        playButtonSound()
        onSkipPrevious?()
    }

    @objc private func nextButtonTapped() {
        playButtonSound()
        onSkipNext?()
    }

    @objc private func endNavButtonTapped() {
        playButtonSound()
        onEndNavigation?()
    }

    private func createProgressCard() -> UIView {
        // Outer container with gray background for visual separation
        let outerCard = UIView()
        outerCard.backgroundColor = UIColor(red: 0.91, green: 0.91, blue: 0.91, alpha: 1.0)  // #E8E8E8
        outerCard.layer.cornerRadius = theme.cornerRadius
        outerCard.layer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner]  // Only top corners rounded
        outerCard.layer.shadowColor = UIColor.black.cgColor
        outerCard.layer.shadowOpacity = 0.12
        outerCard.layer.shadowOffset = CGSize(width: 0, height: 3)
        outerCard.layer.shadowRadius = 8

        // Inner white container for content
        let innerCard = UIView()
        innerCard.backgroundColor = .white
        innerCard.layer.cornerRadius = 12
        innerCard.translatesAutoresizingMaskIntoConstraints = false
        outerCard.addSubview(innerCard)

        NSLayoutConstraint.activate([
            innerCard.topAnchor.constraint(equalTo: outerCard.topAnchor, constant: 10),
            innerCard.leadingAnchor.constraint(equalTo: outerCard.leadingAnchor, constant: 10),
            innerCard.trailingAnchor.constraint(equalTo: outerCard.trailingAnchor, constant: -10),
            innerCard.bottomAnchor.constraint(equalTo: outerCard.bottomAnchor, constant: -10)
        ])

        // Content stack inside the inner white card
        let contentStack = UIStackView()
        contentStack.axis = .vertical
        contentStack.spacing = 10
        contentStack.translatesAutoresizingMaskIntoConstraints = false
        innerCard.addSubview(contentStack)

        NSLayoutConstraint.activate([
            contentStack.topAnchor.constraint(equalTo: innerCard.topAnchor, constant: 12),
            contentStack.leadingAnchor.constraint(equalTo: innerCard.leadingAnchor, constant: 16),
            contentStack.trailingAnchor.constraint(equalTo: innerCard.trailingAnchor, constant: -16),
            contentStack.bottomAnchor.constraint(equalTo: innerCard.bottomAnchor, constant: -16)
        ])

        // === Line 1: [◀] Icon + Waypoint Name [▶] ===
        let line1 = UIStackView()
        line1.axis = .horizontal
        line1.spacing = 12
        line1.alignment = .center

        // Prev button (only if enabled)
        if config.showSkipButtons {
            let prevBtn = createSkipButton(imageName: "chevron.left")
            prevBtn.addTarget(self, action: #selector(prevButtonTapped), for: .touchUpInside)
            prevButton = prevBtn
            line1.addArrangedSubview(prevBtn)
        }

        // Spacer between prev button and icon
        let iconSpacer = UIView()
        iconSpacer.translatesAutoresizingMaskIntoConstraints = false
        iconSpacer.widthAnchor.constraint(equalToConstant: 8).isActive = true
        line1.addArrangedSubview(iconSpacer)

        // Icon container
        let iconBg = UIView()
        iconBg.backgroundColor = theme.primaryColor
        iconBg.layer.cornerRadius = theme.iconSize / 2
        iconBg.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            iconBg.widthAnchor.constraint(equalToConstant: theme.iconSize),
            iconBg.heightAnchor.constraint(equalToConstant: theme.iconSize)
        ])
        iconContainer = iconBg

        let icon = UIImageView()
        icon.image = UIImage(systemName: "flag.fill")
        icon.tintColor = .white
        icon.contentMode = .scaleAspectFit
        icon.backgroundColor = .clear  // Ensure no background on the icon
        icon.translatesAutoresizingMaskIntoConstraints = false
        iconBg.addSubview(icon)
        NSLayoutConstraint.activate([
            icon.centerXAnchor.constraint(equalTo: iconBg.centerXAnchor),
            icon.centerYAnchor.constraint(equalTo: iconBg.centerYAnchor),
            icon.widthAnchor.constraint(equalToConstant: 24),
            icon.heightAnchor.constraint(equalToConstant: 24)
        ])
        iconView = icon

        line1.addArrangedSubview(iconBg)

        // Waypoint name (takes remaining space) - bigger text, closer to icon
        let nameLabel = UILabel()
        nameLabel.text = "Loading..."
        nameLabel.font = .systemFont(ofSize: 18, weight: .medium)
        nameLabel.textColor = theme.textPrimaryColor
        nameLabel.textAlignment = .center
        nameLabel.setContentHuggingPriority(.defaultLow, for: .horizontal)
        waypointNameLabel = nameLabel
        line1.addArrangedSubview(nameLabel)

        // Next button (only if enabled)
        if config.showSkipButtons {
            let nextBtn = createSkipButton(imageName: "chevron.right")
            nextBtn.addTarget(self, action: #selector(nextButtonTapped), for: .touchUpInside)
            nextButton = nextBtn
            line1.addArrangedSubview(nextBtn)
        }

        contentStack.addArrangedSubview(line1)

        // === Line 2: Distance • Time ===
        if config.showDistanceToNext || config.showDurationToNext {
            let distLabel = UILabel()
            distLabel.text = "-- • --"
            distLabel.font = .systemFont(ofSize: 14)
            distLabel.textColor = theme.textSecondaryColor
            distLabel.textAlignment = .center
            distanceTimeLabel = distLabel
            contentStack.addArrangedSubview(distLabel)
        }

        // === Line 3: Progress bar ===
        if config.showProgressBar {
            let progress = UIProgressView(progressViewStyle: .default)
            progress.progress = 0
            progress.progressTintColor = theme.progressBarColor
            progress.trackTintColor = theme.progressBarBackgroundColor
            progress.layer.cornerRadius = 3
            progress.clipsToBounds = true
            progressView = progress
            contentStack.addArrangedSubview(progress)
        }

        // === Line 4: Progress text + Total distance ===
        if config.showWaypointCount || config.showTotalDistance {
            let line4 = UIStackView()
            line4.axis = .horizontal
            line4.distribution = .equalSpacing

            if config.showWaypointCount {
                let progressLabel = UILabel()
                progressLabel.text = "Waypoint 1/1"
                progressLabel.font = .systemFont(ofSize: 13)
                progressLabel.textColor = theme.textSecondaryColor
                progressTextLabel = progressLabel
                line4.addArrangedSubview(progressLabel)
            }

            if config.showTotalDistance {
                let totalLabel = UILabel()
                totalLabel.text = "-- remaining"
                totalLabel.font = .systemFont(ofSize: 13)
                totalLabel.textColor = theme.textSecondaryColor
                totalDistanceLabel = totalLabel
                line4.addArrangedSubview(totalLabel)
            }

            contentStack.addArrangedSubview(line4)
        }

        // === Line 5: ETA ===
        if config.showEta {
            let eta = UILabel()
            eta.text = "ETA --:--"
            eta.font = .systemFont(ofSize: 15, weight: .medium)
            eta.textColor = theme.primaryColor
            eta.textAlignment = .center
            etaLabel = eta
            contentStack.addArrangedSubview(eta)
        }

        // === End Navigation Button ===
        if config.showEndNavigationButton {
            let endBtn = UIButton(type: .system)
            endBtn.setTitle("End Navigation", for: .normal)
            endBtn.setTitleColor(.white, for: .normal)
            endBtn.titleLabel?.font = .systemFont(ofSize: 14, weight: .semibold)
            endBtn.backgroundColor = theme.endButtonColor
            endBtn.layer.cornerRadius = 8
            endBtn.contentEdgeInsets = UIEdgeInsets(top: 12, left: 16, bottom: 12, right: 16)
            endBtn.addTarget(self, action: #selector(endNavButtonTapped), for: .touchUpInside)
            endNavButton = endBtn
            contentStack.addArrangedSubview(endBtn)
        }

        return outerCard
    }

    private func createSkipButton(imageName: String) -> UIButton {
        let button = UIButton(type: .system)
        button.setImage(UIImage(systemName: imageName), for: .normal)
        button.tintColor = theme.primaryColor
        // Darker button background to match Android (#E0E0E0)
        button.backgroundColor = UIColor(red: 0.88, green: 0.88, blue: 0.88, alpha: 1.0)
        button.layer.cornerRadius = 8
        button.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            button.widthAnchor.constraint(equalToConstant: theme.buttonSize),
            button.heightAnchor.constraint(equalToConstant: theme.buttonSize)
        ])
        return button
    }
}

File: android/src/main/kotlin/com/neiladunlop/fluttermapboxnavigation2/utilities/TripProgressOverlay.kt
package com.neiladunlop.fluttermapboxnavigation2.utilities

import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.LayerDrawable
import android.media.AudioManager
import android.media.ToneGenerator
import android.util.Log
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import android.widget.FrameLayout
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.cardview.widget.CardView
import com.neiladunlop.fluttermapboxnavigation2.R
import com.neiladunlop.fluttermapboxnavigation2.activity.NavigationActivity
import com.neiladunlop.fluttermapboxnavigation2.models.TripProgressConfig
import com.neiladunlop.fluttermapboxnavigation2.models.TripProgressData
import com.neiladunlop.fluttermapboxnavigation2.models.TripProgressTheme

/**
 * Displays a trip progress overlay above the navigation info panel.
 *
 * Shows:
 * - Skip prev/next buttons (configurable)
 * - Icon + Next waypoint name + Distance/time to it
 * - Progress bar + "Waypoint X/Y" + Total distance remaining
 * - ETA
 * - End navigation button (configurable)
 *
 * The appearance can be customized via [TripProgressConfig] and [TripProgressTheme].
 */
class TripProgressOverlay(
    private val activity: NavigationActivity,
    private val config: TripProgressConfig = TripProgressConfig.defaults()
) {
    // Theme shortcut
    private val theme: TripProgressTheme get() = config.theme

    private var progressCard: CardView? = null
    private var isVisible = false

    // Callbacks
    var onSkipPrevious: (() -> Unit)? = null
    var onSkipNext: (() -> Unit)? = null
    var onEndNavigation: (() -> Unit)? = null

    // UI elements for updating
    private var iconView: ImageView? = null
    private var iconContainer: FrameLayout? = null
    private var waypointNameView: TextView? = null
    private var distanceTimeView: TextView? = null
    private var progressBar: ProgressBar? = null
    private var progressTextView: TextView? = null
    private var totalDistanceView: TextView? = null
    private var etaView: TextView? = null
    private var prevButton: ImageButton? = null
    private var nextButton: ImageButton? = null
    private var endNavButton: TextView? = null

    // Current state for button enable/disable
    private var currentWaypointIndex = 0
    private var totalWaypoints = 0

    // Bottom margin to position above the info panel
    private val bottomMarginDp = 180

    /**
     * Show the trip progress overlay.
     */
    fun show() {
        if (isVisible) return

        val rootView = activity.findViewById<FrameLayout>(android.R.id.content)
            ?: activity.window.decorView.findViewById<ViewGroup>(android.R.id.content)

        if (rootView == null) {
            Log.e("TripProgressOverlay", "Could not find root view")
            return
        }

        // Create the progress card
        progressCard = createProgressCard()

        // Position at bottom, above the info panel
        val layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        ).apply {
            gravity = Gravity.BOTTOM
            setMargins(dpToPx(12), 0, dpToPx(12), dpToPx(bottomMarginDp))
        }

        // Add to root view
        rootView.addView(progressCard, layoutParams)

        // Animate in
        progressCard?.let { card ->
            card.alpha = 0f
            card.translationY = dpToPx(30).toFloat()
            card.animate()
                .alpha(1f)
                .translationY(0f)
                .setDuration(300)
                .setInterpolator(AccelerateDecelerateInterpolator())
                .start()
        }

        isVisible = true
        Log.d("TripProgressOverlay", "Trip progress overlay shown")
    }

    /**
     * Hide the trip progress overlay.
     */
    fun hide(animated: Boolean = true) {
        progressCard?.let { card ->
            if (animated) {
                card.animate()
                    .alpha(0f)
                    .translationY(dpToPx(30).toFloat())
                    .setDuration(200)
                    .setInterpolator(AccelerateDecelerateInterpolator())
                    .withEndAction {
                        (card.parent as? ViewGroup)?.removeView(card)
                        progressCard = null
                        isVisible = false
                    }
                    .start()
            } else {
                (card.parent as? ViewGroup)?.removeView(card)
                progressCard = null
                isVisible = false
            }
        } ?: run {
            progressCard = null
            isVisible = false
        }
    }

    /**
     * Update the overlay with new progress data.
     */
    fun updateProgress(data: TripProgressData) {
        Log.d("TripProgressOverlay", "updateProgress called: ${data.nextWaypointName}, ${data.progressString}")
        activity.runOnUiThread {
            Log.d("TripProgressOverlay", "Updating UI on main thread, waypointNameView=$waypointNameView")

            currentWaypointIndex = data.currentWaypointIndex
            totalWaypoints = data.totalWaypoints

            // Update icon
            iconView?.setImageResource(getIconResource(data.nextWaypointIconId, data.nextWaypointCategory))
            iconContainer?.background = createCircleDrawable(theme.getCategoryColor(data.nextWaypointCategory))

            // Update waypoint name (truncate if too long)
            val displayName = if (data.nextWaypointName.length > 22) {
                data.nextWaypointName.take(19) + "..."
            } else {
                data.nextWaypointName
            }
            waypointNameView?.text = displayName

            // Update distance and time
            if (config.showDistanceToNext || config.showDurationToNext) {
                val distStr = if (config.showDistanceToNext) data.getFormattedDistanceToNext() else ""
                val timeStr = if (config.showDurationToNext) data.getFormattedDurationToNext() else ""
                val separator = if (config.showDistanceToNext && config.showDurationToNext) " • " else ""
                distanceTimeView?.text = "$distStr$separator$timeStr"
            }

            // Update progress bar
            if (config.showProgressBar) {
                progressBar?.max = 100
                progressBar?.progress = (data.progressFraction * 100).toInt()
            }

            // Update progress text
            if (config.showWaypointCount) {
                progressTextView?.text = data.progressString
            }

            // Update total distance
            if (config.showTotalDistance) {
                totalDistanceView?.text = "${data.getFormattedTotalDistanceRemaining()} remaining"
            }

            // Update ETA
            if (config.showEta) {
                etaView?.text = "ETA ${data.getFormattedEta()}"
            }

            // Update button states
            updateButtonStates()
        }
    }

    private fun updateButtonStates() {
        if (!config.showSkipButtons) return

        // Prev button: disabled if at first waypoint
        val canGoPrev = currentWaypointIndex > 0
        prevButton?.alpha = if (canGoPrev) 1.0f else 0.3f
        prevButton?.isEnabled = canGoPrev

        // Next button: disabled if at last waypoint
        val canGoNext = currentWaypointIndex < totalWaypoints - 1
        nextButton?.alpha = if (canGoNext) 1.0f else 0.3f
        nextButton?.isEnabled = canGoNext
    }

    private fun playButtonSound() {
        if (!config.enableAudioFeedback) return
        try {
            val toneGen = ToneGenerator(AudioManager.STREAM_NOTIFICATION, 50)
            toneGen.startTone(ToneGenerator.TONE_PROP_BEEP, 50)
        } catch (e: Exception) {
            Log.w("TripProgressOverlay", "Could not play button sound: ${e.message}")
        }
    }

    private fun createProgressCard(): CardView {
        val context = activity

        return CardView(context).apply {
            radius = theme.cornerRadius
            cardElevation = dpToPx(6).toFloat()
            setCardBackgroundColor(theme.backgroundColor)
            useCompatPadding = false

            // Main content container
            val contentLayout = LinearLayout(context).apply {
                orientation = LinearLayout.VERTICAL
                setPadding(dpToPx(16), dpToPx(16), dpToPx(16), dpToPx(16))
            }

            // === Line 1: [◀] Icon + Waypoint Name [▶] ===
            val line1 = LinearLayout(context).apply {
                orientation = LinearLayout.HORIZONTAL
                gravity = Gravity.CENTER_VERTICAL
            }

            // Prev button (only if enabled)
            if (config.showSkipButtons) {
                prevButton = createSkipButton(android.R.drawable.ic_media_previous).apply {
                    setOnClickListener {
                        playButtonSound()
                        onSkipPrevious?.invoke()
                    }
                }
                line1.addView(prevButton)
            }

            // Icon container
            iconContainer = FrameLayout(context).apply {
                val size = theme.iconSize
                layoutParams = LinearLayout.LayoutParams(size, size).apply {
                    if (config.showSkipButtons) marginStart = dpToPx(12)
                }
                background = createCircleDrawable(theme.primaryColor)
            }

            iconView = ImageView(context).apply {
                val iconSize = dpToPx(16)
                layoutParams = FrameLayout.LayoutParams(iconSize, iconSize).apply {
                    gravity = Gravity.CENTER
                }
                setImageResource(R.drawable.ic_flag)
                setColorFilter(Color.WHITE)
            }
            iconContainer?.addView(iconView)
            line1.addView(iconContainer)

            // Waypoint name (takes remaining space)
            waypointNameView = TextView(context).apply {
                text = "Loading..."
                setTextSize(TypedValue.COMPLEX_UNIT_SP, 16f)
                setTextColor(theme.textPrimaryColor)
                setPadding(dpToPx(12), 0, dpToPx(12), 0)
                layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f)
                maxLines = 1
                gravity = Gravity.CENTER
            }
            line1.addView(waypointNameView)

            // Next button (only if enabled)
            if (config.showSkipButtons) {
                nextButton = createSkipButton(android.R.drawable.ic_media_next).apply {
                    setOnClickListener {
                        playButtonSound()
                        onSkipNext?.invoke()
                    }
                }
                line1.addView(nextButton)
            }

            contentLayout.addView(line1)

            // === Line 2: Distance • Time ===
            if (config.showDistanceToNext || config.showDurationToNext) {
                distanceTimeView = TextView(context).apply {
                    text = "-- • --"
                    setTextSize(TypedValue.COMPLEX_UNIT_SP, 14f)
                    setTextColor(theme.textSecondaryColor)
                    gravity = Gravity.CENTER
                    setPadding(0, dpToPx(10), 0, 0)
                }
                contentLayout.addView(distanceTimeView)
            }

            // === Line 3: Progress bar ===
            if (config.showProgressBar) {
                progressBar = ProgressBar(context, null, android.R.attr.progressBarStyleHorizontal).apply {
                    layoutParams = LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,
                        dpToPx(6)
                    ).apply {
                        topMargin = dpToPx(10)
                    }
                    max = 100
                    progress = 0
                    progressDrawable = createProgressDrawable()
                }
                contentLayout.addView(progressBar)
            }

            // === Line 4: Progress text + Total distance ===
            if (config.showWaypointCount || config.showTotalDistance) {
                val line4 = LinearLayout(context).apply {
                    orientation = LinearLayout.HORIZONTAL
                    gravity = Gravity.CENTER_VERTICAL
                    setPadding(0, dpToPx(10), 0, 0)
                }

                if (config.showWaypointCount) {
                    progressTextView = TextView(context).apply {
                        text = "Waypoint 1/1"
                        setTextSize(TypedValue.COMPLEX_UNIT_SP, 13f)
                        setTextColor(theme.textSecondaryColor)
                        layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f)
                    }
                    line4.addView(progressTextView)
                }

                if (config.showTotalDistance) {
                    totalDistanceView = TextView(context).apply {
                        text = "-- remaining"
                        setTextSize(TypedValue.COMPLEX_UNIT_SP, 13f)
                        setTextColor(theme.textSecondaryColor)
                        if (!config.showWaypointCount) {
                            layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f)
                        }
                    }
                    line4.addView(totalDistanceView)
                }

                contentLayout.addView(line4)
            }

            // === Line 5: ETA ===
            if (config.showEta) {
                etaView = TextView(context).apply {
                    text = "ETA --:--"
                    setTextSize(TypedValue.COMPLEX_UNIT_SP, 15f)
                    setTextColor(theme.primaryColor)
                    gravity = Gravity.CENTER
                    setPadding(0, dpToPx(10), 0, 0)
                    setTypeface(typeface, android.graphics.Typeface.BOLD)
                }
                contentLayout.addView(etaView)
            }

            // === End Navigation Button ===
            if (config.showEndNavigationButton) {
                endNavButton = TextView(context).apply {
                    text = "End Navigation"
                    setTextSize(TypedValue.COMPLEX_UNIT_SP, 14f)
                    setTextColor(Color.WHITE)
                    gravity = Gravity.CENTER
                    setTypeface(typeface, android.graphics.Typeface.BOLD)
                    setPadding(dpToPx(16), dpToPx(12), dpToPx(16), dpToPx(12))
                    background = createRoundedDrawable(theme.endButtonColor, 8f)
                    layoutParams = LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,
                        LinearLayout.LayoutParams.WRAP_CONTENT
                    ).apply {
                        topMargin = dpToPx(12)
                    }
                    setOnClickListener {
                        playButtonSound()
                        onEndNavigation?.invoke()
                    }
                }
                contentLayout.addView(endNavButton)
            }

            addView(contentLayout)
        }
    }

    private fun createSkipButton(iconRes: Int): ImageButton {
        return ImageButton(activity).apply {
            val size = theme.buttonSize
            layoutParams = LinearLayout.LayoutParams(size, size)
            setImageResource(iconRes)
            setColorFilter(theme.primaryColor, PorterDuff.Mode.SRC_IN)
            background = createRoundedDrawable(theme.buttonBackgroundColor, 8f)
            scaleType = ImageView.ScaleType.CENTER_INSIDE
        }
    }

    private fun createRoundedDrawable(color: Int, radiusDp: Float): GradientDrawable {
        return GradientDrawable().apply {
            shape = GradientDrawable.RECTANGLE
            cornerRadius = dpToPx(radiusDp.toInt()).toFloat()
            setColor(color)
        }
    }

    private fun createProgressDrawable(): GradientDrawable {
        return GradientDrawable().apply {
            shape = GradientDrawable.RECTANGLE
            cornerRadius = dpToPx(3).toFloat()
            setColor(theme.progressBarBackgroundColor)
        }
    }

    private fun createCircleDrawable(color: Int): GradientDrawable {
        return GradientDrawable().apply {
            shape = GradientDrawable.OVAL
            setColor(color)
        }
    }

    private fun getIconResource(iconId: String?, category: String): Int {
        val id = iconId?.lowercase() ?: category.lowercase()

        return when (id) {
            "flag", "checkpoint" -> R.drawable.ic_flag
            "pin", "waypoint" -> R.drawable.ic_pin
            "scenic" -> R.drawable.ic_scenic
            "petrol_station", "petrol", "gas", "fuel" -> R.drawable.ic_petrol_station
            "restaurant", "food" -> R.drawable.ic_restaurant
            "hotel", "accommodation" -> R.drawable.ic_hotel
            "parking" -> R.drawable.ic_parking
            "hospital", "medical" -> R.drawable.ic_hospital
            "police" -> R.drawable.ic_police
            "charging_station", "charging" -> R.drawable.ic_charging_station
            else -> R.drawable.ic_pin
        }
    }

    private fun dpToPx(dp: Int): Int {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            dp.toFloat(),
            activity.resources.displayMetrics
        ).toInt()
    }
}
